File: examples/boschProcess/boschProcess.cpp
#include <geometries/psMakeTrench.hpp>
#include <models/psIsotropicProcess.hpp>
#include <models/psMultiParticleProcess.hpp>
#include <models/psSingleParticleProcess.hpp>
#include <psProcess.hpp>
#include <psUtils.hpp>

using namespace viennaps;
constexpr int D = 2;
using NumericType = double;

void etch(SmartPointer<Domain<NumericType, D>> domain,
          utils::Parameters &params) {
  auto etchModel = SmartPointer<MultiParticleProcess<NumericType, D>>::New();
  etchModel->addNeutralParticle(params.get("neutralStickingProbability"));
  etchModel->addIonParticle(params.get("ionSourceExponent"));
  const NumericType neutralRate = params.get("neutralRate");
  const NumericType ionRate = params.get("ionRate");
  etchModel->setRateFunction(
      [neutralRate, ionRate](const std::vector<NumericType> &fluxes,
                             const Material &material) {
        if (material == Material::Mask)
          return 0.;
        NumericType rate = fluxes[1] * ionRate;
        if (material == Material::Si)
          rate += fluxes[0] * neutralRate;
        return rate;
      });
  Process<NumericType, D>(domain, etchModel, params.get("etchTime")).apply();
}

void punchThrough(SmartPointer<Domain<NumericType, D>> domain,
                  utils::Parameters &params) {
  NumericType depositionThickness = params.get("depositionThickness");
  NumericType gridDelta = params.get("gridDelta");

  // punch through step
  auto depoRemoval = SmartPointer<SingleParticleProcess<NumericType, D>>::New(
      -(depositionThickness + gridDelta / 2.0), 1. /* sticking */,
      params.get("ionSourceExponent"), Material::Mask);
  Process<NumericType, D>(domain, depoRemoval, 1.).apply();
}

void deposit(SmartPointer<Domain<NumericType, D>> domain,
             NumericType depositionThickness) {
  domain->duplicateTopLevelSet(Material::Polymer);
  auto model =
      SmartPointer<IsotropicProcess<NumericType, D>>::New(depositionThickness);
  Process<NumericType, D>(domain, model, 1.).apply();
}

void ash(SmartPointer<Domain<NumericType, D>> domain) {
  domain->removeTopLevelSet();
}

int main(int argc, char **argv) {

  Logger::setLogLevel(LogLevel::INFO);
  omp_set_num_threads(16);

  // Parse the parameters
  utils::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
    return 1;
  }

  // geometry setup
  auto geometry = SmartPointer<Domain<NumericType, D>>::New();
  MakeTrench<NumericType, D>(
      geometry, params.get("gridDelta"), params.get("xExtent"),
      params.get("yExtent"), params.get("trenchWidth"),
      params.get("maskHeight"), 0. /* taper angle */, 0. /* base height */,
      false /* periodic boundary */, true /* create mask */, Material::Si)
      .apply();

  const NumericType depositionThickness = params.get("depositionThickness");
  const int numCycles = params.get<int>("numCycles");

  int n = 0;
  geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");
  etch(geometry, params);
  geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");

  for (int i = 0; i < numCycles; ++i) {
    deposit(geometry, depositionThickness);
    geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");
    punchThrough(geometry, params);
    geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");
    etch(geometry, params);
    geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");
    ash(geometry);
    geometry->saveSurfaceMesh("boschProcess_" + std::to_string(n++) + ".vtp");
  }

  geometry->saveVolumeMesh("boschProcess_final");
}

File: examples/boschProcess/boschProcess.py
from argparse import ArgumentParser

# parse config file name and simulation dimension
parser = ArgumentParser(
    prog="boschProcess",
    description="Run a Bosch process simulation on a trench geometry.",
)
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
args = parser.parse_args()

# switch between 2D and 3D mode
if args.dim == 2:
    print("Running 2D simulation.")
    import viennaps2d as vps
else:
    print("Running 3D simulation.")
    import viennaps3d as vps

params = vps.ReadConfigFile(args.filename)
vps.Logger.setLogLevel(vps.LogLevel.INFO)

geometry = vps.Domain()
vps.MakeTrench(
    domain=geometry,
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
    trenchWidth=params["trenchWidth"],
    trenchDepth=params["maskHeight"],
    taperingAngle=0.0,
    baseHeight=0.0,
    periodicBoundary=False,
    makeMask=True,
    material=vps.Material.Si,
).apply()

# Isoptropic deposition model (emulation of the deposition process)
depoModel = vps.IsotropicProcess(params["depositionThickness"])

depoRemoval = vps.SingleParticleProcess(
    -(params["depositionThickness"] + params["gridDelta"] / 2.0),  # rate
    1.0,  # sticking probability
    params["ionSourceExponent"],  # source exponent
    vps.Material.Mask,  # mask material
)

etchModel = vps.MultiParticleProcess()
etchModel.addNeutralParticle(params["neutralStickingProbability"])
etchModel.addIonParticle(params["ionSourceExponent"])


# Custom rate function for the etch model
def rateFunction(fluxes, material):
    if material == vps.Material.Mask:
        return 0.0
    rate = fluxes[1] * params["ionRate"]
    if material == vps.Material.Si:
        rate += fluxes[0] * params["neutralRate"]
    return rate


etchModel.setRateFunction(rateFunction)

numCycles = int(params["numCycles"])
n = 0

geometry.saveSurfaceMesh("boschProcess_{}".format(n))
n += 1

vps.Process(geometry, etchModel, params["etchTime"]).apply()
geometry.saveSurfaceMesh("boschProcess_{}".format(n))
n += 1

for i in range(numCycles):
    # Deposit a layer of polymer
    geometry.duplicateTopLevelSet(vps.Material.Polymer)
    vps.Process(geometry, depoModel, 1).apply()
    geometry.saveSurfaceMesh("boschProcess_{}".format(n))
    n += 1

    # Remove the polymer layer
    vps.Process(geometry, depoRemoval, 1).apply()
    geometry.saveSurfaceMesh("boschProcess_{}".format(n))
    n += 1

    # Etch the trench
    vps.Process(geometry, etchModel, params["etchTime"]).apply()
    geometry.saveSurfaceMesh("boschProcess_{}".format(n))
    n += 1

    # Ash the polymer
    geometry.removeTopLevelSet()
    geometry.saveSurfaceMesh("boschProcess_{}".format(n))
    n += 1

geometry.saveVolumeMesh("final")
